lib/srv/db/common/packettrace/tracing.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// Trace struct holds all packet details and offers methods to add and save packets.
+type Trace struct {
+	packets    []PacketEntry
+	clock      clockwork.Clock
+	runCommand func(name string, arg ...string) ([]byte, error)
+
+	mx sync.Mutex
+}
+
+// NewTrace initializes a new Trace object.
+func NewTrace(clock clockwork.Clock) *Trace {
+	return &Trace{
+		packets: make([]PacketEntry, 0),
+		clock:   clock,
+		runCommand: func(name string, arg ...string) ([]byte, error) {
+			cmd := exec.Command(name, arg...)
+			out, err := cmd.CombinedOutput()
+			return out, trace.Wrap(err)
+		},
+	}
+}
+
+// AddPacket adds a packet to the trace based on the sender (client or server) and the payload.
+// It automatically handles source and destination addresses based on the sender.
+func (t *Trace) AddPacket(direction Direction, payload []byte) {
+	// Record timestamp
+	timestamp := t.clock.Now()
+
+	// Create a new PacketEntry
+	packet := PacketEntry{
+		Direction: direction,
+		Payload:   payload,
+		Timestamp: timestamp,
+	}
+
+	// Add packet to the trace
+	t.mx.Lock()
+	t.packets = append(t.packets, packet)
+	t.mx.Unlock()
+}
+
+type participant struct {
+	addr      string
+	direction Direction
+}
+
+func (t *Trace) saveOneLinkToPCAP(filename string, port int, sender, receiver participant, packets []PacketEntry) error {
+	buffer := &bytes.Buffer{}
+	for _, packet := range packets {
+		// assign indicator or skip packet.
+		indicator := ""
+		if packet.Direction == sender.direction {
+			indicator = "I"
+		}
+		if packet.Direction == receiver.direction {
+			indicator = "O"
+		}
+		if indicator == "" {
+			continue
+		}
+
+		// Write the timestamp and sender indicator
+		if _, err := fmt.Fprintf(buffer, "%s %s\n", indicator, packet.Timestamp.UTC().Format(time.RFC3339Nano)); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Write the packet data using hex.Dump and add a newline after each packet
+		hexData := hex.Dump(packet.Payload)
+		if _, err := buffer.Write([]byte(hexData + "\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	filenameHex := filename + ".hex"
+
+	err := os.WriteFile(filenameHex, buffer.Bytes(), 0600)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer os.Remove(filenameHex)
+
+	// Invoke text2pcap to convert the hex dump to PCAP format
+	addrPair := fmt.Sprintf("%v,%v", sender.addr, receiver.addr)
+	out, err := t.runCommand(text2pcapBin,
+		// enable inbound/outbound markers
+		"-D",
+		// configure timestamp format
+		"-t", "%Y-%m-%dT%H:%M:%S.%fZ",
+		// ethernet capture encapsulation type
+		"-l", "1",
+		// specify IPv4 addr pair
+		"-4", addrPair,
+		// inbound/outbound ports; also used to hint at packet type; might be worth using "-P <dissector>" instead.
+		"-T", fmt.Sprintf("%v,%v", port, port),
+		filenameHex,
+		filename)
+	if err != nil {
+		return trace.Wrap(err, "error running text2pcap (output: %v)", out)
+	}
+
+	return nil
+}
+
+// SaveToPCAP saves the trace to a single merged pcap file.
```

Comment:
nit: could we document the pre-req?

Reply:
Yes, you are right, this has two runtime dependencies.

Reply:
Documented.